### PR TITLE
Show and reply to messages associated with a post

### PR DESCRIPTION
**Pull request by jasonmule**
This pull request makes the following changes:
- Shows email/ SMS messages related to a post.
- Allow a user to reply to an incoming email/ SMS.

Test these changes by:
- Use with https://github.com/ushahidi/platform-client/pull/146

Fixes ushahidi/platform#443 .

Ping @ushahidi/platform

---
This change is [Reviewable](https://reviewable.io/reviews/ushahidi/platform/999)

### CODE REVIEW

File: application/classes/Ushahidi/DataProvider.php [hunk 1]
```diff
@@ -43,7 +43,7 @@ public function receive($type, $from, $message, $to = NULL, $title = NULL, $data
 		$usecase = service('factory.usecase')->get('messages', 'receive');
 		try
 		{
-			$usecase->setPayload(compact(['type', 'from', 'message', 'to', 'title', 'data_provider_message_id', 'data_provider', 'contact_type']))
+			$message_id = $usecase->setPayload(compact(['type', 'from', 'message', 'to', 'title', 'data_provider_message_id', 'data_provider', 'contact_type']))
```
Comment (rjmackay) [inline application/classes/Ushahidi/DataProvider.php]: Do we need this change? if so can we update the docblock to with a correct @return

File: plugins/smssync/classes/Controller/Sms/Smssync.php [hunk 2]
```diff
@@ -91,7 +98,7 @@ private function _incoming()
 		// Remove Non-Numeric characters because that's what the DB has
 		$to = preg_replace("/[^0-9,.]/", "", $this->request->post('sent_to'));
 
-		$from = preg_replace("/[^0-9,.]/", "", $from);
+		$from = preg_replace("/[^0-9A-Za-z,.]/", "", $from);
```
Comment (rjmackay) [inline plugins/smssync/classes/Controller/Sms/Smssync.php]: Might be best to make this even more forgiving.. If we can get alpha chars in phone numbers, we can probably get all of unicode letters.

File: src/Core/Usecase/Message/CreateMessage.php [hunk 3]
```diff
@@ -21,6 +21,12 @@ protected function getEntity()
 	{
 		$entity = parent::getEntity();
 
+		if (! empty($this->payload['parent_id'])) {
```
Comment (rjmackay) [inline src/Core/Usecase/Message/CreateMessage.php]: Needs a comment.

**Comment by rjmackay** [comment]
Looks good. Nice and simple mostly :)

File: application/classes/Ushahidi/Validator/Contact/Update.php [hunk 4]
```diff
@@ -69,18 +69,8 @@ public function valid_contact($contact, $data, $validation)
...
-		{
-			// Remove all non-digit characters from the number
-			$number = preg_replace('/\D+/', '', $contact);
-
-			if (strlen($number) == 0)
-			{
-				$validation->error('contact', 'invalid_phone', [$contact]);
-			}
-		}
+		// There's no easy to validate phone numbers. Even just accepting numbers
```
Comment (rjmackay) [inline application/classes/Ushahidi/Validator/Contact/Update.php]: @jasonmule moved this over from msheria too. We hit an issue where SMSSync was forwarding messages with a phone number like "Safaricom" .. however this probably removes the validation we wanted creating contacts from the UI.. Any ideas for a solution to both?

Comment (jasonmule) [inline application/classes/Ushahidi/Validator/Contact/Update.php]: We probably shouldn't be using a regex to validate phone numbers. Maybe use https://github.com/googlei18n/libphonenumber but that still won't validate "Safaricom". For now, we could use a separate validator for the receive use case?

File: plugins/smssync/classes/Controller/Sms/Smssync.php (hunk 2 above)
Comment (jasonmule) [inline plugins/smssync/classes/Controller/Sms/Smssync.php]: Maybe we should limit this to alphanumeric senders? i.e 1 - 11 alphanumeric characters including a space.

File: application/classes/Ushahidi/DataProvider.php (hunk 1 above)
Comment (jasonmule) [inline application/classes/Ushahidi/DataProvider.php]: Proabably not.

**Comment by rjmackay** [comment]
Reviewed 7 of 11 files at r1, 1 of 1 files at r2, 6 of 6 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/ushahidi/platform/999)*

**Comment by rjmackay** [comment]
Good to go :shipit: 

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/ushahidi/platform/999#-:-KFuZqmJ8y4k0NYMN8Te:1178353558)*

**Pull request merged**